torch/ao/sparsity/scheduler/base_scheduler.py

```diff
@@ -0,0 +1,129 @@
+
+from torch.ao.sparsity import BaseSparsifier
+
+from functools import wraps
+import weakref
+
+class BaseScheduler(object):
+
+    def __init__(self, sparsifier, last_epoch=-1, verbose=False):
+
+        # Attach sparsifier
+        if not isinstance(sparsifier, BaseSparsifier):
+            raise TypeError('{} is not an instance of torch.ao.sparsity.BaseSparsifier'.format(
+                type(sparsifier).__name__))
+        self.sparsifier = sparsifier
+
+        # Initialize epoch and base sparsity levels
+
+        self.base_sl = [group['sparsity_level'] for group in sparsifier.module_groups]
+        self.last_epoch = last_epoch
+
+        # Following https://github.com/pytorch/pytorch/issues/20124
```

Comment:
The logic here is hard to follow. Is this the only way to implement this?

Reply:
I put as many comments in here. But in general, I agree this is a little convoluted. I will change the first comment to make sure it is more clear.